airflow/models/taskinstance.py

```diff
@@ -2128,6 +2135,28 @@ def set_duration(self) -> None:
             self.duration = None
         self.log.debug("Task Duration set to %s", self.duration)
 
+    @provide_session
+    def _record_task_map_for_downstreams(self, value: Any, *, session: Session = NEW_SESSION) -> None:
+        if not self.task.has_mapped_dependants():
+            return
+        if not isinstance(value, collections.abc.Collection):
+            return  # TODO: Error if the pushed value is not mappable?
+        session.query(TaskMap).filter_by(
+            dag_id=self.dag_id,
+            task_id=self.task_id,
+            run_id=self.run_id,
+            map_index=self.map_index,
+        ).delete()
```

Comment:
If we are doing session.merge we don't _have to_ delete it as SQLA would handle doing the right INSERT/UPDATE itself.